src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,134 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private T[] grow(int oldSize) {
+        int newSize = oldSize + oldSize / 2;
+        T[] newElementData = (T[]) new Object[newSize];
+        System.arraycopy(elementData, 0, newElementData, 0, elementData.length);
+        return newElementData;
+    }
+
+    public void checkNonExistentOrNegativePosition(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add element on this position");
+        }
+    }
+
+    public void checkIndexForAddElement(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add element on this position");
+        }
+    }
+
+    public void addElementInsideArrayByIndex(T value, int index) {
+        T[] newElementData = (T[]) new Object[size + 1];
+        if ((size + 1) > elementData.length) {
+            grow(size + 1);
+        }
+        if (index == 0) {
+            System.arraycopy(elementData, 0, newElementData, 1, size);
+            elementData = newElementData;
+            elementData[index] = value;
+            size++;
+        } else {
+            System.arraycopy(elementData, 0, newElementData, 0, index);
+            System.arraycopy(elementData, index, newElementData, index + 1,
+                    size - index);
+            elementData = newElementData;
+            elementData[index] = value;
+            size++;
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAddElement(index);
+        if (index == elementData.length) {
+            elementData = grow(size);
+        }
+        if (index >= 0 && index < size) {
+            addElementInsideArrayByIndex(value, index);
+            return;
+        }
+        elementData[index] = value;
+        size++;
```

Comment:
do you know that this if condition is always true and last two lines never execute? Use one System.arraycopy instead of if-block

Reply:
here's the line: 
`System.arraycopy(elements, index, elements, index + 1, size - index);`

Reply:
It isn't true when index = 0 and size = 0, and it isn't true when index = size... If I'll put this line instead of "if" - I have to change other methods... so I don't sure if it matters...

Reply:
I understood what you meant. I have changed the logic of this method by adding only one System.arraycopy() and I have changed "if" logic...